servicex/topcp/topcp.py

```diff
@@ -0,0 +1,137 @@
+from ..query_core import QueryStringGenerator
+
+
+@pydantic.dataclasses.dataclass
+class TopCPQuery(QueryStringGenerator):
+    yaml_tag = "!TopCP"
+    default_codegen = "topcp"
+
+    reco_yaml: Optional[Union[Path, str]] = None
+    """Path to the reco.yaml"""
+    parton_yaml: Optional[Union[Path, str]] = None
+    """Path to the parton.yaml"""
+    particle_yaml: Optional[Union[Path, str]] = None
+    """Path to the particle.yaml"""
+    max_events: Optional[int] = -1
+    """Number of events to process"""
+    parton: Optional[bool] = False
+    """Toggles the parton-level analysis"""
+    particle: Optional[bool] = False
+    """Toggles the particle-level analysis"""
+    no_reco: Optional[bool] = False
+    """Toggles off the detector-level analysis"""
+    no_systematics: Optional[bool] = True
+    """Toggles off the computation of systematics"""
+    no_filter: Optional[bool] = False
+    """Save all events regardless of analysis filters (still saves the decision)"""
+
+    @pydantic.model_validator(mode="after")
+    def check_reco_yaml(self):
+        if self.reco_yaml is None and self.no_reco is False:
+            raise ValueError("reco is enabled but reco.yaml is missing!")
+        return self
+
+    @pydantic.model_validator(mode="after")
+    def no_input_yaml(self):
+        if (
+            self.reco_yaml is None
+            and self.parton_yaml is None
+            and self.particle_yaml is None
+        ):
+            raise ValueError("No yaml provided!")
+        return self
+
+    @pydantic.model_validator(mode="after")
+    def no_parton_yaml(self):
+        if self.parton_yaml is None and self.parton is True:
+            raise ValueError("parton is set to True but no parton.yaml provided!")
+        return self
+
+    @pydantic.model_validator(mode="after")
+    def no_paricle_yaml(self):
+        if self.particle_yaml is None and self.particle is True:
+            raise ValueError("particle is set to True but no particle.yaml provided!")
+        return self
+
+    @pydantic.model_validator(mode="after")
+    def no_run(self):
+        if self.no_reco is True and self.particle is False and self.parton is False:
+            raise ValueError("Wrong configuration - no reco, no particle, no parton!")
+        return self
+
+    def generate_selection_string(self):
+        import yaml
+        import json
+
+        recoYaml = None
+        if self.reco_yaml:
+            with open(self.reco_yaml, "r") as reco_file:
+                recoYaml = yaml.safe_load(reco_file)
+
+        partonYaml = None
+        if self.parton_yaml:
+            with open(self.parton_yaml, "r") as parton_file:
+                partonYaml = yaml.safe_load(parton_file)
+
+        particleYaml = None
+        if self.particle_yaml:
+            with open(self.particle_yaml, "r") as particle_file:
+                particleYaml = yaml.safe_load(particle_file)
+
+        query = {
+            "RecoYAML": recoYaml,
```

Comment:
Q on this -
* maybe it would be easier just to call it "reco", "parton", "particle" ?
* I guess we should assume that if the user provides a YAML file they wish to run that particular type of analysis? so by default if someone provides a particle YAML then RunParticle should be True?

Reply:
> * maybe it would be easier just to call it "reco", "parton", "particle" ?

I agree. Let me also change camel case field names to snake case.

> * I guess we should assume that if the user provides a YAML file they wish to run that particular type of analysis? so by default if someone provides a particle YAML then RunParticle should be True?

I'm following the default values of TopCPToolkit except `no_systematics`. I'm open to change the defaults but I guess it's common to have all `reco.yaml`, `parton.yaml` and `particle.yaml` in the same directory and mostly run `reco.yaml` only.

Reply:
Oh? How does TopCP determine which one(s) to run?

It just seems to me that there is unlikely to be any situation where someone provides a particle-level YAML but doesn't expect the particle analysis to be run by default... ?

Reply:
That's fair - TopCP doesn't take each yaml input file. A user provides a directory which contains all yaml files instead. I think your proposal makes more sense as ServiceX query. Let me drop options like `run_parton`.